### ISSUE TITLE
Benchmark tool for Read performance

### ISSUE THREAD

**Issue by sdg002** [Azure/azure-cosmos-dotnet-v3#1240]
**Is your feature request related to a problem? Please describe.**
The benchmark tool under the Samples folder is very useful to get Write statistics. It would be very useful if this tool can be extended to perform query operations. E.g. Query using the document id.

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos.Samples/Tools/Benchmark/Program.cs

**Describe the solution you'd like**
Add an option to the existing benchmark tool to support Read only operations and print out the performance. E.g. x queries per sec, N RUs/sec

**Describe alternatives you've considered**
Writing my own stuff.

**Additional context**
NA

**Comment by j82w** [comment]
What exactly is the goal you are trying to achieve with these statistics?

**Comment by sdg002** [comment]
The Benchmark tool does a good job of giving us some insights into the Write performance.
I was hoping to get similar data w.r.t Read performance.

Unless, you think that Read performance is not of any material value. I am happy to change my mind. Please let me know.

Thank you.

**Comment by j82w** [comment]
The reason I asked is the benchmarks tests are to help detect performance regressions in the SDK. Since the goal is to test the SDK it was decided to mock out the end point to get the most reliable testing. If the goal is to get actual performance of an end to end scenario then the current benchmarks are not going to do that.

**Comment by sdg002** [comment]
Thanks for replying. The goal was precisely to benchmark how good/bad does a CosmosDB perform. If I were to provision 800 RU/s as against 400 RU/s , then how much do the numbers change under the hood for my documents.   
Example - https://www.youtube.com/watch?feature=youtu.be&v=bQBeTeYUrR8&app=desktop


End  2 end scenario test is something that would also be of interest - but that is out of scope in this discussion.

**Comment by j82w** [comment]
There is no way to validate 800 RU/s vs 400 RU/s using the current benchmark tests as they all use a mocked endpoint an instantly return the results. You won't see the throttling that would be caused by using a real Cosmos DB instance. Also there is no latency from the Cosmos DB service or networking. I think your best option is to create your own Benchmark project with document size and partitioning scheme your application would use.

1. Follow the [performance guide](https://docs.microsoft.com/en-us/azure/cosmos-db/performance-tips).
2. The cost to read a [1 KB item is 1 Request Unit](https://docs.microsoft.com/en-us/azure/cosmos-db/request-units) (or 1 RU). This can be used to estimate how many RU's you will need for the desired number of operations.
3. There is batch and bulk support if through put is needed more than latency.
4. The current range of RU's most likely will still have a single physical partition. To get multiple partitions you will need to scale to much larger values. The number changes as the service becomes more optimized and hardware becomes more powerful. 
5. For query the partition key is very important. You should try to chose a value that will correlate the data you plan on querying to the same partition key when possible to avoid doing a cross partition query. 
6. [Query performance](https://docs.microsoft.com/en-us/azure/cosmos-db/sql-api-query-metrics) will change a significantly based on the query. The best way to get a benchmark is with a realistic items and the expected query. For example doing a query using a field that is index lookup will be faster than query using a field that requires a full scan.

**Comment by ealsur** [comment]
@j82w the Benchmark project @sdg002 refers to is not a mocked project. It is a real tool to measure real performance against a real account.

@sdg002 will show you the read information too:
![image](https://user-images.githubusercontent.com/1633401/75795792-a44f7580-5d27-11ea-86af-b8d58a86b8f8.png)

Since writes consume more RU/s than reads, they are the most common RU/s consumption scenario, 

Is that information not good enough to compare performance?

**Comment by j82w** [comment]
@ealsur is there anything blocking adding a query sample?

**Comment by ealsur** [comment]
The thing is, the tool is not a sample. There is a json file you can customize with your sample payload, and the tool will try to measure how many point operations are possible from a throughput perspective.

We can add queries, but the question there is, does it make sense a scenario where users are concurrently calling the same query and want to know how many concurrent queries they can run? If so, then yes. And they should be able to customize the query too (not sure we can nail a generic query good for everyone) like we allow the payload customization.

**Comment by sdg002** [comment]
@ealsur 
You are right. It does not make sense to query with the same parameters repetitively. 

However, I can provide you with scenarios from my current sphere of work where I felt the need for a benchmarking.

**Application scenario**
We are building an app which is capturing biometrics from end users. 

- 10,000 users
- 6 months of biometric data per user. 
- 1 record per user per day

Structure of the JSON is as follows:

```
{
    customerid:"10002",
    timestamp:"2020-04-06T10:36:38",
    weight: 123.12,
    bloodpressure: 34.12
}

```

**How much throughput can I get from Cosmos?**
Assuming, I have populated Cosmos DB with synthetic data , I would like to know how much throughput can I get out of Cosmos at X RUs/second of provisioning.  Obviously, we will have to bump up the RU if we are not getting the desired results.

**Other important questions that business would need an answer to**
- Are there any improvements if I change the partitioning strategy?
- What happens when the user count goes up to 20,000? 

This is where a benchmarking tool will give us some clues. I should be able to point the tool to my Cosmos DB instance  and specify parameterized queries and optionally specify partition keys.

**Posisble solution**
- The benchmarking tool should be able to read a parameteried query from a configuration file
- The parameters of the query could be read from a CSV file, the name of which is also specified through configuration
- The responsiblity of the tool user to configure the query and provide the CSV file

Thank you

**Comment by ealsur** [comment]
The problem is that the answer to your question:
> How much throughput can I get from Cosmos?
It really is a matter of volume of operations and type of operations. If you see the Calculator (https://cosmos.azure.com/capacitycalculator/) the RU/s you need to provision varies depending on the type and volume of each operation.

In your scenario, queries, are you going to be running queries concurrently and you are interested in knowing how many queries can you execute in parallel?

**Comment by sdg002** [comment]
I would like to know the following by doing a 5 minute stress test using a simple tool like the **Cosmos DB bench-marking** tool or **Apache JMeter**.

**Given**
A cosmos db provisioned at X ru/s
A JSON payload shown in the post above
A partitioning key choice (e.g. I might add a Country Code or a State Code)
The database is pre-populated with N thousand objects (could be 1000, 5000, 10000 depending on the projected growth of our application in the coming months)

**Determine the following**
Throughput (requests per second)
95 percentile latency of the requests

I like the simplicity of the Cosmos DB bench-marking tool. However, for my current problem I am going to proceed with **Apache JMeter**.  I intend to invoke the REST API of the Cosmos DB and feed in parameters from a CSV file.

Thank you.

**Comment by ealsur** [comment]
Makes sense. We can leave this open for whenever we have time, I think it should be a different tool, like a Query Benchmark Tool, to avoid polluting the base benchmark code with ifs.

**Comment by sdg002** [comment]
Agree. On second thoughts - it might make more sense for the product team to provide tutorials/guidance on how to use a tool like JMeter (since VS load test has been deprecated) as opposed to building a fresh new tool.
